### ISSUE TITLE
Problems during training.

### ISSUE THREAD

**Issue by XuyangBai** [HuguesTHOMAS/KPConv#15]
Hi, Thanks for your sharing. I have tried your code on my own dataset but the I found that initially everything goes well but after several epochs the training suddenly broke up ( accuracy becomes 1 and the loss becomes 0 ) I use tf 1.12.0 and the cuda version is 9.0, cudnn version is 7.1.4
```bash
# conda list | grep tensorflow
tensorflow-estimator      1.13.0                     py_0    anaconda
tensorflow-gpu            1.12.0                   pypi_0    pypi
tensorflow-tensorboard    0.4.0                    pypi_0    pypi
```
Have you met this kind of problem? Another potential problem is that sometimes the training takes 4400 MB GPU memory (see from `nvidia-smi`), but sometimes it takes more than 7000 MB ( and I do not change the batch size and network architecture) I am pretty confused about these problems. Could you give me some advice?

**Comment by HuguesTHOMAS** [comment]
Hi @XuyangBai,

The first error is quite strange and I never encountered such behavior on my datasets. It is very unlikely that the loss really became zero if you use correct augmentation strategies. It seems more like a bug, but it will be difficult to help you without reproducing your experiments with your dataset.

The second one could be explained by your dataset. The GPU memory that you see from `nvidia-smi` is the memory taken by the tensors at runtime. It thus depends on the input size. If you use the same network parameters, but with a different dataset, with denser point cloud for example, this memory will be larger. It is strange that you get different values GPU memory for the same dataset, but might be explained by the nature of your dataset and your implementation of the dataset class. 

What does your data look like? Real or artificial point clouds? Indoor outdour scenes? Objects?

**Comment by XuyangBai** [comment]
Hi @HuguesTHOMAS ,

I also think the first error may because some bugs of my implementation. I just want to check whether it is because the tensorflow version. BTW, what's the situation when using TF 1.13 and CUDA 10 ?

For the second one, I mean that for some experiments, the GPU memory is always 4400MB and for others is always 7000+MB. It is really strange.  But I check the `training.txt` in result fold and found the memory showed in that file is similar.

**Comment by XuyangBai** [comment]
Oh Sorry I think I find the reason of second problem.  I forget the dropout. It seems when I use dropout = 0.5, the GPU memory is around 4400 MB while using dropout = 1, the GPU memory is 7000MB. 

Sorry for the bothering.

**Comment by HuguesTHOMAS** [comment]
Hi @XuyangBai,

If you look at the code, the dropout variable is extremely important in the implementation, because the network use it to know if you use it for training or for test.

If you use a dropout < 0.99, the network is in training configuration, and if you use a dropout = 1, the network is in test configuration. This is a trick that I used to avoid creating a 'training/test' boolean placeholder, and that I never correct. 

It will be corrected in the next month (I currently don't have any time to spend on the code). Until then, you should not use dropout = 1 when training, as the variables will not be updated by gradient back propagation in that case. If you have dropout blocks and don't want to use them, just remove them or use dropout = 0.98 and they will be insignificant.

Best,
Hugues

**Comment by XuyangBai** [comment]
Thanks

**Comment by XuyangBai** [comment]
Thanks a lot for your reply :)

**Comment by XuyangBai** [comment]
Hi @HuguesTHOMAS 

I reopen this issue because I go through the code to check out the bug but I am still facing the first problem that `training will broke up (loss become zero)`. Actually, everything goes well for environment `CUDA 9, TF 1.12.0`. But when I tried to run the code on my RTX 2080Ti with `CUDA 10, TF 1.12.0` (build from source), the problem will appear. And I printed the input value and variables for when the model broken and found the input value is OK but **all the variables**  all NaN. 

I have tried to add 1e-12 to operation like `tf.sqrt` to avoid infinite number but the problem is still there.

Another thing need to mention is that sometimes the training and validation both 'broke up' while sometimes the training seems correct but the validation is broken up, like the curve below.
<img width="370" alt="屏幕快照 2019-08-03 下午7 23 55" src="https://user-images.githubusercontent.com/15166089/62411372-4c7f4d80-b624-11e9-8c24-56a231170427.png">

Have you ever met such problems like the variables become all NaN?  Thank you so much.

**Comment by nejcd** [comment]
Hi @XuyangBai, I have noticed similar behaviour as you have described. I have not been able to debug it as it occurs randomly.
As you have closed the issue, have you found some solution?

**Comment by XuyangBai** [comment]
Hi @nejcd , I didn't find the solution, so I just change my environment to `CUDA9 TF 1.12.0` and everything goes well. There might be some bugs in CUDA10.

**Comment by HuguesTHOMAS** [comment]
Hi,

I had some time to dig into this problem and it seems that CUDA10 is not working correctly with RTX 2080Ti GPUs. Here is what I found:

**Tested configurations**
* CUDA9-TF1.12 / GTX 1080ti => No bug
* CUDA10-TF1.13 / GTX 1080ti => No bug
* CUDA9-TF1.12 / RTX 2080ti => No bug
* CUDA10-TF1.13 / RTX 2080ti => Bug appears only in this configuration


**Origin of the bug**
I tracked down the NaN values in my code and found that they appear after a tf.matmul operation:
https://github.com/HuguesTHOMAS/KPConv/blob/5f9cecae72c7a22cc4247b7852eee707ecab8fcd/kernels/convolution_ops.py#L240

Before the appearance of NaN I noticed some weird value higher than 1e10. If you print the two matrix that are multiplied and the result matrix, you will see that the result is completly false. This seems to be caused by tensorflow of CUDA internal bug. At some point one of this mistake lead to a value so high that it becomes NaN and the network crashes.

For now I would just advise to avoid using CUDA10 with a RTX 2080ti.

**Comment by miiller** [comment]
I ported the model to Keras layers and tried training it on a Tesla V100 GPU (CUDA10.2:tf2.0) with the result of also getting NaN values after some epochs. After changing the KP influence from guassian to linear everything worked fine, so I would assume the issue lies in the gradient computation for the Gaussian influence, although increasing the epsilon from 1e-9 to 1e-6 did not resolve the problem. But the linear influence works just fine and in my case leads to good results with higher computational efficiency.

**Comment by longmalongma** [comment]
>  after a tf.matmul operation:

Thanks for your great work,I have met this problem for a long time,i want to konw the version of python.

**Comment by HuguesTHOMAS** [comment]
If I remember correctly, the python version was 3.5 or 3.6. IF you are willing to switch libraries, a newer implementation has been released in PyTorch.

**Comment by Arjun-NA** [comment]
Just to mention my experience:

I got NaN when I used Tensorflow 1.15, cuda 10 cudnn 7.6.5
with some specific configurations only 
with GPU : NVIDIA Tesla P100

**Comment by densechen** [comment]
I use TF 1.15 and also occur NaN. You can solve this via reducing the batch size with 2.

**Comment by wuqianliang** [comment]
Nice.
